sycl/test-e2e/VirtualFunctions/misc/group-barrier.cpp

```diff
@@ -0,0 +1,140 @@
+// FIXME: replace unsupported with an aspect check once we have it
+// UNSUPPORTED: cuda, hip, acc
+//
+// REQUIRES: aspect-usm_shared_allocations
+//
+// Fails with UR_RESULT_ERROR_PROGRAM_LINK_FAILURE. SPIR-V files produced by
+// SYCL_DUMP_IMAGES can be linked just fine (using llvm-spirv -r + llvm-link),
+// so it seems to be a problem on IGC side.
+// Reported in https://github.com/intel/llvm/issues/15068
+// On CPU it segfaults within the kernel that performs virtual function call.
+// https://github.com/intel/llvm/issues/15080
+// XFAIL: gpu, cpu
+//
+// This test checks that group operations (barrier in this case) work correctly
+// inside virtual functions.
+//
+// RUN: %{build} -o %t.out %helper-includes
+// RUN: %{run} %t.out
+
+#include <sycl/detail/core.hpp>
+#include <sycl/group_barrier.hpp>
+#include <sycl/usm.hpp>
+
+#include "helpers.hpp"
+
+#include <iostream>
+#include <numeric>
+
+namespace oneapi = sycl::ext::oneapi::experimental;
+
+class BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *, sycl::nd_item<1>) = 0;
+};
+
+class SumOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *LocalData, sycl::nd_item<1> It) {
+    LocalData[It.get_local_id()] += It.get_local_id();
+    sycl::group_barrier(It.get_group());
```

Comment:
I don't exactly share the concern. I.e. `apply` could have been a regular function which can also be blindly copy-pasted and called from a non-convergent/non-uniform context resulting in the very same UB.

Reply:
Maybe using `apply(int *, sycl::group)` would be a better pattern? `group` in arguments is what the spec uses for such interfaces.

Reply:
Thinking about it more. I can't just pass `group` only, because I need local IDs which aren't available in `group`. And `nd_item` already includes `group`, so passing them both together would be a bit weird>

I suppose that we should assume that if `nd_item` is passed, then some group operations can be performed.

Reply:
There is `group::get_local_id` in core SYCL.

Reply:
> There is `group::get_local_id` in core SYCL.

Didn't know that! Switched to use `group` instead of `nd_item` in 0ea83a84e5b5f9b7f2d4216a4c8129cafdbf93bb

Host reference calculation has also been fixed by that commit: I've verified it on CPU (test passes there with internal newer version of OCL CPU RT).